src/rtc.rs

```diff
@@ -0,0 +1,729 @@
+        // Enter initialization mode
+        self.reg.isr.modify(|_, w| w.init().set_bit());
+        while self.reg.isr.read().initf().bit_is_clear() {}
+
+        self.reg.tr.reset();
+        self.reg.dr.reset();
+
+        // Exit initialization mode
+        self.reg.isr.modify(|_, w| w.init().clear_bit());
+    }
+
+    /// Returns `None` if the calendar is uninitialized
+    fn calendar_initialized(&self) -> Option<()> {
+        match self.reg.isr.read().inits().bit() {
+            true => Some(()),
+            false => None,
+        }
+    }
+
+    /// Wait for initialization or shift to complete
+    fn wait_for_sync(&self) {
+        while self.reg.isr.read().rsf().bit_is_clear() {}
+    }
+
+    /// Calendar Date
+    ///
+    /// Returns `None` if the calendar has not been initialized
+    pub fn date(&self) -> Option<NaiveDate> {
+        self.calendar_initialized()?;
+        self.wait_for_sync();
+        let data = self.reg.dr.read();
+        let year = 2000 + i32(data.yt().bits()) * 10 + i32(data.yu().bits());
+        let month = data.mt().bits() as u8 * 10 + data.mu().bits();
+        let day = data.dt().bits() * 10 + data.du().bits();
+        NaiveDate::from_ymd_opt(year, u32(month), u32(day))
+    }
+
+    /// Current Time
+    ///
+    /// Returns `None` if the calendar has not been initialized
+    pub fn time(&self) -> Option<NaiveTime> {
+        self.calendar_initialized()?;
+        self.wait_for_sync();
+        let data = self.reg.tr.read();
+        let mut hour = data.ht().bits() * 10 + data.hu().bits();
+        if data.pm().bit_is_set() {
+            hour += 12;
+        }
+        let minute = data.mnt().bits() * 10 + data.mnu().bits();
+        let second = data.st().bits() * 10 + data.su().bits();
+        let micro = self.ss_to_us(self.reg.ssr.read().ss().bits());
+        NaiveTime::from_hms_micro_opt(
+            u32(hour),
+            u32(minute),
+            u32(second),
+            micro,
+        )
+    }
+
+    /// Calendar Date and Time
+    ///
+    /// Returns `None` if the calendar has not been initialized
+    pub fn date_time(&self) -> Option<NaiveDateTime> {
+        let date = self.date()?;
+        let time = self.time()?;
+        Some(date.and_time(time))
+    }
+
+    /// Returns the fraction of seconds that have occurred since the last second tick
+    ///
+    /// The precision of this value depends on the value of the synchronous prescale divider
+    /// (prediv_s) which depends on the frequency of the RTC clock. E.g. with a 32,768 Hz
+    /// crystal this value has a resolution of 1/256 of a second.
+    pub fn subseconds(&self) -> Option<f32> {
+        self.calendar_initialized()?;
+        self.wait_for_sync();
+        let ss = f32(self.reg.ssr.read().ss().bits());
+        let prediv_s = f32(self.reg.prer.read().prediv_s().bits());
+        Some((prediv_s - ss) / (prediv_s + 1.0))
+    }
+
+    fn ss_to_us(&self, ss: u16) -> u32 {
```

Comment:
It looks like this method returns milliseconds, not microseconds

Reply:
Yeah, I remember now. I was trying to avoid overflowing u32 but also avoid needing u64 divide since `__aeabi_uldivmod` is many kilobytes. I'll have to think about what to do here.

Reply:
Ok, yes that is tricky. You could split the multiply to take advantage of as much u32 precision as possible, like so:

```rust
assert!(ss <= prediv_s); // See RM0433 Rev 7 46.6.10, shift operations not supported

debug_assert!(prediv_s < 32_768);
let tens_us = ((prediv_s - ss) * 100_000) / (prediv_s + 1);

tens_us * 10
```

Reply:
You figured that out faster than me, should've come back and checked. I'll add an assert for the possibility of ss > prediv_s.